From 777a31a4274bb80570f8339c8ba19024f6b98d3f Mon Sep 17 00:00:00 2001
From: Nicolas LOUBOUTIN <nicolas.louboutin@st.com>
Date: Fri, 6 Aug 2021 17:22:05 +0200
Subject: [PATCH] Android: enable AVB mechanism only for eMMC

Error to access RPMB when booting with microSD card, bypass AVB verification in this case

Signed-off-by: Nicolas LOUBOUTIN <nicolas.louboutin@st.com>
---
 include/configs/stm32mp1.h | 31 ++++++++++++++++++-------------
 1 file changed, 18 insertions(+), 13 deletions(-)

diff --git a/include/configs/stm32mp1.h b/include/configs/stm32mp1.h
index a6559c75..63319d89 100644
--- a/include/configs/stm32mp1.h
+++ b/include/configs/stm32mp1.h
@@ -175,7 +175,7 @@
 		"then " \
 		   "mmc read ${splashimage} ${splash_start} ${splash_size};" \
 		   "cls; bmp display ${splashimage} m m;" \
-		"fi\0" \
+		"fi;\0" \
 	"android_mmc_fdt=" \
 		"if part start mmc ${devnum} dt_${suffix} dt_start &&" \
 		   "part size mmc ${devnum} dt_${suffix} dt_size;" \
@@ -184,7 +184,7 @@
 		   "dtimg getindex ${dtimg_addr} ${board_id} ${board_rev}" \
 		     " dt_index;" \
 		   "dtimg start ${dtimg_addr} ${dt_index} fdt_addr_r;" \
-		"fi\0" \
+		"fi;\0" \
 	"android_mmc_kernel=" \
 		"if part start mmc ${devnum} boot_${suffix} boot_start && " \
 		   "part size mmc ${devnum} boot_${suffix} boot_size;" \
@@ -193,28 +193,33 @@
 		   "env set bootargs ${android_bootargs} " \
 		     "androidboot.serialno=${serial#} " \
 		     "androidboot.slot_suffix=_${suffix};" \
-		"fi\0" \
+		"fi;\0" \
 	"avb_verify=avb init ${devnum}; avb verify _${suffix};\0" \
 	"android_mmc_boot=" \
-		"if run avb_verify;" \
-		"then " \
-			"echo AVB verification OK. Continue boot;" \
+		"if test \"${devnum}\" = \"1\"; then " \
+			"if run avb_verify; then " \
+				"echo AVB verification OK. Continue boot;" \
+				"mmc dev ${devnum};" \
+				"run android_mmc_splash;" \
+				"run android_mmc_fdt;" \
+				"run android_mmc_kernel;" \
+				"bootm ${kernel_addr_r} ${kernel_addr_r} ${fdt_addr_r};" \
+			" else " \
+				"echo AVB verification failed;" \
+				"exit;" \
+			" fi; " \
+		"else " \
+			"echo bypass AVB verification for microSD card;" \
 			"mmc dev ${devnum};" \
 			"run android_mmc_splash;" \
 			"run android_mmc_fdt;" \
 			"run android_mmc_kernel;" \
 			"bootm ${kernel_addr_r} ${kernel_addr_r} ${fdt_addr_r};" \
-		"else " \
-			"echo AVB verification failed;" \
-			"exit;" \
-		"fi\0" \
+		"fi;\0" \
 	"bootcmd_android=" \
 		"env set mmc_boot run android_mmc_boot;" \
 		"run bootcmd_stm32mp\0"
 
-
-
-
 #else
 #define STM32MP_ANDROID
 #endif/* CONFIG_CMD_DTIMG */
-- 
2.17.1

